File: android/src/main/java/com/lightningkite/khrysalis/observables/WriteAddedObservableProperty.shared.kt
package com.lightningkite.khrysalis.observables

import com.lightningkite.khrysalis.Box
import com.lightningkite.khrysalis.escaping
import io.reactivex.Observable

class WriteAddedObservableProperty<A>(
    val basedOn: ObservableProperty<A>,
    val onWrite: @escaping() (A) -> Unit
) : MutableObservableProperty<A>() {
    override var value: A
        get() = basedOn.value
        set(value) {
            onWrite(value)
        }
    override val onChange: Observable<Box<A>> get() = basedOn.onChange
    override fun update() {
        //Do nothing
    }
}

fun <T> ObservableProperty<T>.withWrite(
    onWrite: @escaping() (T) -> Unit
): MutableObservableProperty<T> {
    return WriteAddedObservableProperty<T>(this, onWrite)
}

File: ios/Khrysalis/com/lightningkite/khrysalis/observables/WriteAddedObservableProperty.shared.swift
//Package: com.lightningkite.khrysalis.observables
//Converted using Khrysalis2

import Foundation
import RxSwift
import RxRelay



public class WriteAddedObservableProperty<A>: MutableObservableProperty<A> {
    
    public var basedOn: ObservableProperty<A>
    public var onWrite:  (A) -> Void
    
    override public var value: A {
        get {
            return basedOn.value
        }
        set(value) {
            onWrite(value)
        }
    }
    override public var onChange: Observable<A> {
        get {
            return basedOn.onChange
        }
    }
    
    override public func update() -> Void {
    }
    
    public init(basedOn: ObservableProperty<A>, onWrite: @escaping (A) -> Void) {
        self.basedOn = basedOn
        self.onWrite = onWrite
        super.init()
    }
    convenience public init(_ basedOn: ObservableProperty<A>, _ onWrite: @escaping (A) -> Void) {
        self.init(basedOn: basedOn, onWrite: onWrite)
    }
}
 
 

extension ObservableProperty {
    public func withWrite(onWrite: @escaping (T) -> Void) -> MutableObservableProperty<T> {
        return WriteAddedObservableProperty<T>(self, onWrite)
    }
}
 
